core/spec/models/spree/promotion_spec.rb

```diff
@@ -773,4 +773,42 @@
       expect(order.adjustment_total).to eq -10
     end
   end
+
+  describe '#duplicate' do
+    let(:promotion) { build(:promotion, code: "123") }
+
+    subject { promotion.duplicate }
+
+    it 'returns a new promotion' do
+      new_promo = subject
+      expect(new_promo).to_not eq promotion
+    end
+
+    context "with rules" do
+      let!(:promotion) { create(:promotion_with_item_total_rule, code: '123') }
+
+      it 'has valid duplicated rules' do
+        new_promo_rule = subject.rules.first
+        promo_rule = promotion.rules.first
+
+        expect(new_promo_rule.type).to eq promo_rule.type
+        expect(new_promo_rule.preferences).to eq promo_rule.preferences
+        expect(new_promo_rule).to be_valid
+      end
+    end
+
+    context "with actions" do
+      let!(:promotion) { create(:promotion_with_order_adjustment, code: '123') }
+
+      it 'has valid duplicated actions' do
+        new_promo_action = subject.actions.first
```

Comment:
We probably need to setup the actions to ensure duplicate creates all of them especially if we are relying on `first` here.